test/IISIntegration.FunctionalTests/Inprocess/AppOfflineTests.cs

```diff
@@ -0,0 +1,130 @@
+using Microsoft.AspNetCore.Server.IntegrationTesting;
+using Microsoft.AspNetCore.Testing.xunit;
+using Microsoft.Extensions.Logging;
+using Xunit;
+
+namespace IISIntegration.FunctionalTests.Inprocess
+{
+    [SkipIfIISExpressSchemaMissingInProcess]
+    public class AppOfflineTests : IISFunctionalTestBase
+    {
+        [ConditionalFact]
+        public async Task AppOfflineDroppedWhileSiteIsDown_SiteReturns503()
+        {
+            var deploymentParameters = Helpers.GetBaseDeploymentParameters();
+
+            var deploymentResult = await DeployAsync(deploymentParameters);
+
+            AddAppOffline(deploymentResult.DeploymentResult.ContentRoot);
+
+            await AssertAppOffline(deploymentResult);
+        }
+
+        [ConditionalFact]
+        public async Task AppOfflineDroppedWhileSiteIsDown_CustomResponse()
+        {
+            var expectedResponse = "The app is offline.";
+
+            var deploymentParameters = Helpers.GetBaseDeploymentParameters();
+
+            var deploymentResult = await DeployAsync(deploymentParameters);
+
+            AddAppOffline(deploymentResult.DeploymentResult.ContentRoot, expectedResponse);
+
+            await AssertAppOffline(deploymentResult, expectedResponse);
+        }
+
+        [ConditionalFact]
+        public async Task AppOfflineDroppedWhileSiteRunning_SiteShutsDown()
+        {
+            var deploymentResult = await AssertStarts();
+
+            AddAppOffline(deploymentResult.DeploymentResult.ContentRoot);
+
+            await AssertStopsProcess(deploymentResult);
+        }
+
+        [ConditionalFact]
+        public async Task AppOfflineDropped_CanRemoveAppOfflineAfterAdding()
+        {
+            var deploymentResult = await AssertStarts();
+
+            AddAppOffline(deploymentResult.DeploymentResult.ContentRoot);
+
+            await AssertStopsProcess(deploymentResult);
+
+            RemoveAppOffline(deploymentResult.DeploymentResult.ContentRoot);
+        }
+
+        private void AddAppOffline(string appPath, string content = "")
+        {
+            File.WriteAllText(Path.Combine(appPath, "app_offline.htm"), content);
+        }
+
+        private void RemoveAppOffline(string appPath)
+        {
+            RetryHelper.RetryOperation(
+                () => File.Delete(Path.Combine(appPath, "app_offline.htm")),
+                e => Logger.LogError($"Failed to remove app_offline : {e.Message}"),
+                retryCount: 3,
+                retryDelayMilliseconds: 100);
+        }
+
+        private async Task AssertAppOffline(IISDeploymentResult deploymentResult, string expectedResponse = "")
+        {
+            var response = await deploymentResult.RetryingHttpClient.GetAsync("HelloWorld");
+
+            for (var i = 0; response.IsSuccessStatusCode && i < 5; i++)
+            {
+                // Keep retrying until app_offline is present.
+                response = await deploymentResult.RetryingHttpClient.GetAsync("HelloWorld");
```

Comment:
The client is already retrying, do you need to retry even more?
How long does it take?

Reply:
yeah no need for retrying.